### PR TITLE
Add delete, recover and banned Project logic in backend

### LINKED ISSUE #1438: Add delete, ban and recover project logic in backend
* Add recover, ban and delete project in backend
* Change the permission to work with the project, taking into roles in the organization

### PR DESCRIPTION
**Pull request by Arrgentum**
## Whats added:
- Add deleted, banned and recovery Project logic
- Add new permission ```BAN```
- Correct tests

## Issue
This pull request close #1438

### CODE REVIEW

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ProjectService.kt [hunk 1]
```diff
@@ -77,8 +120,24 @@ class ProjectService(
...
+
+    /**
+     * @param name
+     * @param organizationName
+     * @return project
      */
-    fun findByNameAndOrganizationName(name: String, organizationName: String) = projectRepository.findByNameAndOrganizationName(name, organizationName)
+    fun findByNameAndOrganizationName(name: String, organizationName: String): Project? =
+            findByNameAndOrganizationNameWithoutStatus(name, organizationName)
+                ?.let {
```
Comment (nulls) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ProjectService.kt:134]:
```suggestion
                ?.takeIf { it.organization.status == OrganizationStatus.CREATED }
```

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ProjectService.kt [hunk 2]
```diff
@@ -51,6 +48,52 @@ class ProjectService(
...
+            projectRepository.save(it)
+        }
+
+    /**
+     * Mark organization [project] as deleted
+     *
+     * @param project an [project] to delete
+     * @return deleted organization
+     */
+    fun deleteProject(project: Project): Project =
```
Comment (nulls) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ProjectService.kt:73]: `@Transactional`

**Review by nulls — APPROVED** [review]

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ProjectService.kt (hunk 2 above)
Comment (Arrgentum) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ProjectService.kt:73]: I looked, they don't seem to change the values of several tables at once, then I can delete in this file. Or is it better to do it everywhere?

Comment (Arrgentum) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ProjectService.kt:73]: ok

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ProjectService.kt (hunk 1 above)
Comment (Arrgentum) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/service/ProjectService.kt:134]: Done

**Review by Cheshiriks — APPROVED** [review]

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/security/ProjectPermissionEvaluator.kt [hunk 3]
```diff
@@ -82,16 +100,19 @@ class ProjectPermissionEvaluator(
...
 
-    private fun hasDeleteAccess(userId: Long?, projectRole: Role): Boolean =
-            hasBanAccess(userId, projectRole) || userId?.let { projectRole == Role.OWNER } ?: false
+    private fun hasDeleteAccess(userId: Long?, projectRole: Role, organzationRole: Role): Boolean =
+            hasBanAccess(userId, projectRole, organzationRole) ||
+                    userId?.let { projectRole == Role.OWNER || organzationRole == Role.OWNER } ?: false
 
-    private fun hasBanAccess(userId: Long?, projectRole: Role): Boolean = userId?.let { projectRole == Role.SUPER_ADMIN } ?: false
+    private fun hasBanAccess(userId: Long?, projectRole: Role, organzationRole: Role): Boolean = userId?.let {
+        projectRole == Role.SUPER_ADMIN || organzationRole == Role.SUPER_ADMIN
```
Comment (sanyavertolet) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/security/ProjectPermissionEvaluator.kt:114]: It is never so btw as far as `Role.SUPER_ADMIN` cannot be assigned on organization/project level. It is a special role that is defined on a higher level.

**Review by sanyavertolet — APPROVED** [review]

**Review by akuleshov7 — APPROVED** [review]

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/security/ProjectPermissionEvaluator.kt (hunk 3 above)
Comment (Arrgentum) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/security/ProjectPermissionEvaluator.kt]: Done

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ProjectController.kt [hunk 4]
```diff
@@ -247,34 +250,61 @@ class ProjectController(
...
+        description = "Change status existing project by its name.",
+    )
+    @Parameters(
+        Parameter(name = "organizationName", `in` = ParameterIn.PATH, description = "name of an organization", required = true),
+        Parameter(name = "projectName", `in` = ParameterIn.PATH, description = "name of a project", required = true),
+        Parameter(name = "status", `in` = ParameterIn.QUERY, description = "type of status being set", required = true),
     )
-    @ApiResponse(responseCode = "200", description = "Successfully deleted a project.")
-    @ApiResponse(responseCode = "403", description = "Not enough permission for project deletion.")
+    @ApiResponse(responseCode = "200", description = "Successfully change status a project.")
```
[the remark this answers is not in the document]
Comment (Arrgentum) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ProjectController.kt]: Done

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ProjectController.kt [hunk 5]
```diff
@@ -247,34 +250,61 @@ class ProjectController(
...
+    @PostMapping("/{organizationName}/{projectName}/change-status")
     @RequiresAuthorizationSourceHeader
     @PreAuthorize("permitAll()")
     @Operation(
-        method = "DELETE",
-        summary = "Delete a project.",
-        description = "Delete a project.",
+        method = "POST",
+        summary = "Change status existing project.",
+        description = "Change status existing project by its name.",
```
[the remark this answers is not in the document]
Comment (Arrgentum) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ProjectController.kt]: Done

File: save-backend/src/test/kotlin/com/saveourtool/save/backend/security/ProjectPermissionEvaluatorTest.kt [hunk 6]
```diff
@@ -123,6 +125,7 @@ class ProjectPermissionEvaluatorTest {
         )
     }
 
+
```
Comment (nulls) [inline save-backend/src/test/kotlin/com/saveourtool/save/backend/security/ProjectPermissionEvaluatorTest.kt:128]: extra line

Comment (Arrgentum) [reply save-backend/src/test/kotlin/com/saveourtool/save/backend/security/ProjectPermissionEvaluatorTest.kt]: Done

**Pull request merged**